File: Search/binary_search_step.py
import csv

def load_dataset(filename):
    data = []
    with open(filename, mode='r') as file:
        reader = csv.reader(file)
        for row in reader:
            data.append((int(row[0]), row[1]))
    return data

def record_step(step_list, filename):
    with open(filename, 'w') as file:
        for step in step_list:
            file.write(step + '\n')

def binary_search_step(data, target, output_filename):
    left = 0
    right = len(data) - 1
    steps = []

    while left <= right:
        mid = (left + right) // 2
        number, text = data[mid]
        steps.append(f"{mid + 1}: {number}/{text}")

        if number == target:
            record_step(steps, output_filename)
            return

        if number < target:
            left = mid + 1
        else:
            right = mid - 1

    steps.append("-1")
    record_step(steps, output_filename)
    return

target_value = 1018802776
dataset_filename = "dataset/quick_sort_result/quick_sort_1000.csv"
output_filename = f"dataset/binary_search_result/binary_search_step_{target_value}.txt"

data = load_dataset(dataset_filename)
binary_search_step(data, target_value, output_filename)

File: Search/BinarySearchStep.java
import java.io.*;
import java.util.*;

public class BinarySearchStep {

    public static List<int[]> loadDataset(String filename) {
        List<int[]> data = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;

            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",");
                int number = Integer.parseInt(parts[0]);
                String text = parts[1];
                data.add(new int[]{number, text.hashCode()});
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return data;
    }

    public static void recordStep(List<String> steps, String outputFilename) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFilename))) {
            for (String step : steps) {
                writer.write(step);
                writer.newLine();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void binarySearchStep(List<int[]> data, int target, String outputFilename) {
        int left = 0;
        int right = data.size() - 1;
        List<String> steps = new ArrayList<>();

        while (left <= right) {
            int mid = (left + right) / 2;
            int number = data.get(mid)[0];
            steps.add((mid + 1) + ": " + number);

            if (number == target) {
                recordStep(steps, outputFilename);
                return;
            }

            if (number < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        steps.add("-1");
        recordStep(steps, outputFilename);
    }

    public static void main(String[] args) {
        int targetValue = 4925744;
        String datasetFilename = "dataset/quick_sort_result/quick_sort_1000.csv";
        String outputFilename = "dataset/binary_search_result/binary_search_step_" + targetValue + ".txt";

        List<int[]> data = loadDataset(datasetFilename);
        binarySearchStep(data, targetValue, outputFilename);
    }
}
